service/controller/resource/instance/delete.go

```diff
@@ -2,8 +2,17 @@ package instance
 
 import (
 	"context"
+
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
 )
 
 func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.ClientFactory.RemoveAllClients(cr)
```

Comment:
why remove the clients when removing a cluster?

Reply:
I mean, maybe two clusters are using the same credential?

Reply:
Wouldn't this also be wrong point in reconciliation loop? The clients are needed later in the reconciliation flow and would get created again while possibly still allow finalizer get dropped in the last one so this wouldn't get called again.

Could the clients be cached for some time and then get cleaned up when unused for 30min or so?

Reply:
> I mean, maybe two clusters are using the same credential?

Yup, true. I'm adding cluster tracking, so it's removed only when the last one is removed.

> The clients are needed later in the reconciliation flow and would get created again while possibly still allow finalizer get dropped in the last one so this wouldn't get called again.

Not sure I understand this. You're saying clients are used after `EnsureDeleted`? In that case, yeah those wouldn't get deleted.

> Could the clients be cached for some time and then get cleaned up when unused for 30min or so?

We can do something like this.

Reply:
> Not sure I understand this. You're saying clients are used after EnsureDeleted? In that case, yeah those wouldn't get deleted.

Yeah. Concrete example:

- These are resources executed over single reconciliation period: https://github.com/giantswarm/azure-operator/blob/master/service/controller/cluster.go#L566-L584
- After `instance` resource there are still:
  - `endpointsResource`
  - `vpnResource`
  - `vpnconnectionResource`
- At least `vpnconnectionResource` needs client, although currently it's not using the one from factory, but maybe it should?

Reply:
> At least vpnconnectionResource needs client, although currently it's not using the one from factory, but maybe it should?

IMO all resources should get clients from the factory. Maybe just not all of them in this pull request :)

Reply:
The factory is now using go-cache for storing clients, and the clients are cached for 30 minutes.

Reply:
More precisely, the cache duration is refreshed on every access. So if some client is not used for 30 minutes, it's removed from cache.